numba/np/ufunc/gufunc.py

```diff
@@ -60,75 +119,77 @@ def _rebuild(cls, py_func, signature, identity, cache, is_dynamic,
     def __repr__(self):
         return f"<numba._GUFunc '{self.__name__}'>"
 
+    def _install_type(self, typingctx=None):
+        """Constructs and installs a typing class for a gufunc object in the
+        input typing context.  If no typing context is given, then
+        _install_type() installs into the typing context of the
+        dispatcher object (should be same default context used by
+        jit() and njit()).
+        """
+        if typingctx is None:
+            typingctx = self._dispatcher.targetdescr.typing_context
+        _ty_cls = type('GUFuncTyping_' + self.__name__,
+                       (AbstractTemplate,),
+                       dict(key=self, generic=self._type_me))
+        typingctx.insert_user_function(self, _ty_cls)
+
     def add(self, fty):
         self.gufunc_builder.add(fty)
 
     def build_ufunc(self):
         self.ufunc = self.gufunc_builder.build_ufunc()
         return self
 
-    def disable_compile(self):
+    def expected_ndims(self):
+        parsed_sig = parse_signature(self.gufunc_builder.signature)
+        return tuple(map(len, parsed_sig[0])) + tuple(map(len, parsed_sig[1]))
+
+    def _type_me(self, argtys, kws):
         """
-        Disable the compilation of new signatures at call time.
+        Implement AbstractTemplate.generic() for the typing class
+        built by gufunc._install_type().
+
+        Return the call-site signature after either validating the
+        element-wise signature or compiling for it.
         """
-        # If disabling compilation then there must be at least one signature
-        assert len(self.gufunc_builder._sigs) > 0
-        self._frozen = True
+        assert not kws
+        ufunc = self.ufunc
+        _handle_inputs_result = npydecl.Numpy_rules_ufunc._handle_inputs(
+            ufunc, argtys, kws)
+        ewise_types, _, _, _ = _handle_inputs_result
+        sig, _ = self.find_ewise_function(ewise_types)
+
+        if sig is None:
+            # Matching element-wise signature was not found; must
+            # compile.
+            if self._frozen:
+                raise TypeError("cannot call %s with types %s"
+                                % (self, argtys))
+            self._compile_for_argtys(ewise_types)
+            # double check to ensure there is a match
+            sig, _ = self.find_ewise_function(ewise_types)
+            assert sig is not None
```

Comment:
This assertion is triggered when no output argument is supplied. It will print error ilke:

```python
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'GUFunc._type_me': File: numba/np/ufunc/gufunc.py: Line 147.
    With argument(s): '(array(int64, 3d, C))':
   Rejected as the implementation raised a specific error:
     AssertionError:
  raised from /path/to/dev/numba/numba/np/ufunc/gufunc.py:171

During: resolving callee type: Function(<numba._GUFunc 'gufunc'>)
During: typing of call at /Users/siu/dev/numba/chk.py (11)


File "chk.py", line 11:
def jit_func(x, res):
    res[:] = gufunc(x)
    ^
````

I suggest raising a `TypingError` with messages to help user.